dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java

```diff
@@ -136,28 +133,27 @@ private LinkageChecker(
 
   /**
    * Searches the classpath for linkage errors.
-   * 
-   * @return {@link SymbolProblem}s found in the class path and referencing classes
+   *
+   * @return {@link LinkageProblem}s found in the class path and referencing classes
    * @throws IOException I/O error reading files in the classpath
    */
-  public ImmutableSetMultimap<SymbolProblem, ClassFile> findSymbolProblems() throws IOException {
-    // Having Problem in key will dedup SymbolProblems
-    ImmutableSetMultimap.Builder<SymbolProblem, ClassFile> problemToClass =
-        ImmutableSetMultimap.builder();
-    
+  public ImmutableSet<LinkageProblem> findLinkageProblems() throws IOException {
+    ImmutableSet.Builder<LinkageProblem> problemToClass = ImmutableSet.builder();
+
+    // This sourceClassFile is a source of references to other symbols.
     for (ClassFile classFile : symbolReferences.getClassFiles()) {
       ImmutableSet<ClassSymbol> classSymbols = symbolReferences.getClassSymbols(classFile);
-
       for (ClassSymbol classSymbol : classSymbols) {
         if (classSymbol instanceof SuperClassSymbol) {
```

Comment:
if blocks testing subtypes suggests there's a missed opportunity to use polymorphism. Could the logic of each if block be pushed into a method in ClassSymbol and its subtypes instead?